File: pso-cpp/src/Point.cpp
#include "Point.h"
#include <functional>
#include <vector>

Point::Point(std::vector<int> startPos, std::vector<double> startVelocity)
    :
    _position(startPos),
    _velocityVector(startVelocity),
    _personalBest(startPos)
    {}

void Point::evalPoint(std::function<double(const std::vector<int>&)> funcToMinimize)
{
    auto currentGrade = funcToMinimize(_position);
    if (currentGrade < _grade) 
    {
        _personalBest = _position;
        _grade = currentGrade;
    }
}

void Point::updatePoisiton(void)
{
    for(auto i = 0; i < _position.size(); i++)
    {
        _position[i] += _velocityVector[i];
    }
}

void Point::updateVelocity(float alpha, float beta, float epsilon1, float epsilon2, std::vector<int> globalBest)
{
    for(auto i = 0; i < _velocityVector.size(); i++)
    {
        _velocityVector[i] += (alpha * epsilon1 * (globalBest[i] - _position[i]) + 
                               beta * epsilon2 * (_personalBest[i] - _position[i]));
    }
}

File: pso-rust/src/point.rs

pub struct Point 
{
    position: Vec<f64>,
    velocityVector: Vec<f64>,
    personalBest: Vec<f64>,
    grade: f64,
}

impl Point 
{
    pub fn new(startPos: Vec<f64>, startVelocity: Vec<f64>) -> Self 
    {
        Point {
            position: startPos.clone(),
            velocityVector: startVelocity,
            personalBest: startPos,
            grade: f64::MAX,
        }
    }    

    pub fn evalPoint(&mut self, func: &dyn Fn(&[f64]) -> f64)
    {
        let currentGrade = func(&self.position);

        if currentGrade < self.grade 
        {
            self.personalBest = self.position.clone();
            self.grade = currentGrade;    
        }
    }

    pub fn updatePosition(&mut self)
    {
        for (pos, vel) in self.position.iter_mut().zip(self.velocityVector.iter())
        {
            *pos += *vel;
        }
    }

    pub fn updateVelocity(&mut self, alpha: f64, beta: f64, epsilon1: f64, epsilon2: f64, globalBest: &Vec<f64>)
    {
        for i in 0..self.velocityVector.len()
        {
            self.velocityVector[i] += alpha * epsilon1 * (globalBest[i] - self.position[i]) + 
                                      beta * epsilon2 * (self.personalBest[i] - self.position[i]);
        }
    }

    pub fn enforceBounds(&mut self, bounds: (f64, f64))
    {
        for i in 0..self.velocityVector.len()
        {
            self.position[i] = bounds.0.max(bounds.1.min(self.position[i]));
        }
    }
}  